criu/files-reg.c

```diff
@@ -1380,12 +1381,290 @@ static bool should_check_size(int flags)
+	Elf64_Phdr *program_header, *program_header_end;
+	Elf64_Nhdr *note_header, *note_header_end;
+
+	/* 
+	 * If the file doesn't have atleast 1 program header entry, it definitely can't
+	 * have a build-id.
+	 */
+	if (!file_header->e_phnum) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+	file_header_end = (size_t) fd_status->st_size + (size_t) file_header;
+
+	program_header = (Elf64_Phdr *) (file_header->e_phoff + (char *) file_header);
+	if (program_header <= (Elf64_Phdr *) file_header) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	program_header_end = (Elf64_Phdr *) (file_header_end - sizeof(Elf64_Phdr));
+	num_iterations = file_header->e_phnum + 1;
+
+	/* 
+	 * If the file has a build-id, it will be in the PT_NOTE program header 
+	 * entry AKA the note sections.
+	 */
+	while (num_iterations-- && program_header <= program_header_end &&
+			program_header->p_type != PT_NOTE)
+		program_header++;
+
+	if (!num_iterations || program_header >= program_header_end) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	note_header = (Elf64_Nhdr *) (program_header->p_offset + (char *) file_header);
+	if (note_header <= (Elf64_Nhdr *) file_header) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	note_header_end = (Elf64_Nhdr *) (file_header_end - sizeof(Elf64_Nhdr));
+	num_iterations = 500;
+
+	/* The note type for the build-id is NT_GNU_BUILD_ID. */
+	while (num_iterations-- && note_header <= note_header_end &&
+			note_header->n_type != NT_GNU_BUILD_ID)
+		note_header = (Elf64_Nhdr *) ((char *) note_header + sizeof(Elf64_Nhdr) +
+						note_header->n_namesz + note_header->n_descsz);
+
+	if (!num_iterations || note_header >= note_header_end) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	if (note_header->n_descsz <= 0 || note_header->n_descsz > 512) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	size = note_header->n_descsz;
+	note_header = (Elf64_Nhdr *) ((char *) note_header + sizeof(Elf64_Nhdr) +
+					note_header->n_namesz);
+	note_header_end = (Elf64_Nhdr *) (file_header_end - size);
+	if (note_header <= (Elf64_Nhdr *) file_header || note_header > note_header_end) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	*build_id = (unsigned char *) xmalloc(size);
+	if (!*build_id) {
+		munmap(file_header, fd_status->st_size);
+		return -1;
+	}
+
+	memcpy(*build_id, (void *) note_header, size);
+
+	munmap(file_header, fd_status->st_size);
+	return size;
+}
+
+/*
+ * Finds the build-id of the file by checking if the file is an ELF file
+ * and then calling either the 32-bit or the 64-bit function as necessary.
+ * Returns the number of bytes of the build-id if it could be
+ * obtained, else -1.
+ */
+static int get_build_id(const int fd, const struct stat *fd_status,
+				unsigned char **build_id)
+{
+	char *start_addr;
+
+	if (fd_status->st_size < 5)
+		return -1;
+	start_addr = (char *) mmap(0, fd_status->st_size,
```

Comment:
The file can be very big. Are you sure that we need to map the whole file? Maybe we can avoid this somehow? And If we can't, we need to check the elf magic before mapping a file.

Reply:
This shouldn't be a problem since demand paging is used right? mmap is only anticipatory and it will fault and demand the required pages from the file system only when needed.

Reply:
On x32 systems, the address space size is only 4Gb. CRIU has to work on arm32.

Reply:
Instead of mapping the file to check for the magic number, I'll read the first 5 bytes and only if it's an ELF file I'll map it?

Reply:
Even in this case, you need to think about how to handle a case when you can't map a file due to its size.

Reply:
One possible solution is that if a file is greater than 3 GB (4GB minus 1 GB for the system reserved address spaces), regardless of which validation method the user has opted for, I read the first 1024 bytes (Some large number of bytes) and use the checksum of these bytes for validation.

Another solution would be to always mmap at most 3 GB (If the file is greater, I'll just pretend it's only 3 GB and try finding the build-ID or checksum).

Reply:
I'm not sure which would be a better approach but I think it's the second one.
@0x7f454c46 Is there a better solution?

Reply:
Well, yes, I'd use some "big-enough" chunk to mmap(). Let's say, 1Mb (define it as a constant in one place in case there's need to change it) and when you found where are the section headers, mmap() only them (with this 1 Mb window), then mmap() another 1Mb on top of previous while looking for build-id.

Reply:
@0x7f454c46 That seems unnecessary though, since the build-id exists within the first hundreds of bytes (I tested this by printing the location, in bytes, of the build-id relative to the start of each file and ran the ZDTM test suite) so I can just mmap() the first 1 MB and if the build-id doesn't exist within this, it's highly probable that it doesn't exist at all and I will call the checksum method instead. Wouldn't this be better?

Reply:
Well, for computing the checksum you still will need this window (1 Mb or whatever sensable limit).

Reply:
@0x7f454c46 Yes but for finding the build-id this is unnecessary since only the first chunk is needed.